File: backend/app/core/upgrades/stat_modifiers/dash_speed_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class DashSpeed1(UpgradeImplementation):
    id = 'dash_speed_1'
    name = 'Swift Escape'
    description = '+15% dash speed'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.15
        appliers.modifiers.add_multiplier('player', 'dashSpeed', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/dash_speed_1.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const DashSpeed1Def: UpgradeDef = {
  id: "dash_speed_1",
  name: "Swift Escape",
  description: "+15% dash speed",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 10,
  targetClass: UpgradeTargetID.Player,
  fieldInTargetClass: UpgradeStatID.DashSpeed,
  value: 0.15,
  isMultiplier: true,
  stackable: true,
  maxStacks: 3,
  dependentOn: [{ ids: ["dash_ability"] }],
}

export class DashSpeed1 extends Upgrade {}
